src/corefx/System.Globalization.Native/pal_collation.c

```diff
@@ -24,34 +24,18 @@ const int32_t CompareOptionsIgnoreWidth = 0x10;
 // Windows and Unix platforms. The nonalphanumeric symbols will come after alphanumeric
 // characters on Windows, but before on Unix.
 // Since locale - specific string sort order can change from one version of Windows to the next,
-// there is no reason to guarantee string sort order between Windows and ICU. Thus trying to 
+// there is no reason to guarantee string sort order between Windows and ICU. Thus trying to
 // change ICU's default behavior here isn't really justified unless someone has a strong reason
 // for !StringSort to behave differently.
 
-typedef std::map<int32_t, UCollator*> TCollatorMap;
-typedef std::pair<int32_t, UCollator*> TCollatorMapPair;
+typedef struct { UChar* items; size_t size; } UCharList;
 
-/*
- * For increased performance, we cache the UCollator objects for a locale and
- * share them across threads. This is safe (and supported in ICU) if we ensure
- * multiple threads are only ever dealing with const UCollators.
- */
-typedef struct _sort_handle
+int TreeComparer(const void* left, const void* right)
 {
-    UCollator* regular;
-    TCollatorMap collatorsPerOption;
-    pthread_mutex_t collatorsLockObject;
-
-    _sort_handle() : regular(nullptr)
-    {
-        int result = pthread_mutex_init(&collatorsLockObject, NULL);
-        if (result != 0)
-        {
-            assert(false && "Unexpected pthread_mutex_init return value.");
-        }
-    }
-
-} SortHandle;
+    const TCollatorMap* leftMap = left;
+    const TCollatorMap* rightMap = right;
+    return leftMap->key - rightMap->key;
```

Comment:
This can return wrong result in a general case. (Not sure for the range of keys that are possible here.)

Change this to correct by construction pattern?

Reply:
I think this is implemented correctly and in accord with the documentation https://linux.die.net/man/3/tsearch. No issues found so far.

Reply:
E.g. when left key is `0x7FFFFFFF` and the right key key is `-1`, this will return incorrect result.

The correct way to implement this looks like this: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Int32.cs#L51

This issue causes corner case bugs that are not easy to find.